Realm/RLMObservation.mm

```diff
@@ -0,0 +1,732 @@
+        }
+    });
+
+    block();
+
+    for (auto const& change : reverse(changes)) {
+        change.info->didChange(change.property, NSKeyValueChangeRemoval, change.indexes);
+    }
+    for (auto info : reverse(invalidated)) {
+        info->didChange(RLMInvalidatedKey);
+    }
+
+    realm.group->set_cascade_notification_handler(nullptr);
+}
+
+namespace {
+class TransactLogHandler {
+    // For advance_read() and friends, we need to batch up the change
+    // notifications and send them all at once, since all of the changes are
+    // applied atomically. ObserverState tracks everything needed to send the
+    // notifications for a single observed row.
+    struct ObserverState {
+        size_t table;
+        size_t row;
+        RLMObservationInfo *info;
+
+        // Change information for a single property/column
+        struct change {
+            bool changed = false;
+            bool multipleLinkviewChanges = false;
+            NSKeyValueChange linkviewChangeKind = NSKeyValueChangeSetting;
+            NSMutableIndexSet *linkviewChangeIndexes = nil;
+        };
+        std::vector<change> changes;
+
+        // Get the change info for the given column, creating it if needed
+        change& getChange(size_t i) {
+            if (changes.size() <= i) {
+                changes.resize(std::max(changes.size() * 2, i + 1));
+            }
+            return changes[i];
+        }
+
+        // Loop over the columns which were changed
+        template<typename Func>
+        void forEach(Func&& f) const {
+            for (size_t i = 0; i < changes.size(); ++i) {
+                auto const& change = changes[i];
+                if (change.changed) {
+                    f(i, change);
+                }
+            }
+        }
+
+        // Simple lexographic ordering
+        friend bool operator<(ObserverState const& lft, ObserverState const& rgt) {
+            return std::tie(lft.table, lft.row) < std::tie(rgt.table, rgt.row);
+        }
+    };
+
+    size_t currentTable = 0;
+    std::vector<ObserverState> observers;
+    std::vector<RLMObservationInfo *> invalidated;
+
+    ObserverState::change *activeLinkList = nullptr;
+
+    // Find all observed objects in the given object schema and build up the
+    // array of observers to notify from them
+    void findObservers(NSArray *schema) {
+        for (RLMObjectSchema *objectSchema in schema) {
+            for (auto info : objectSchema->_observedObjects) {
+                auto const& row = info->getRow();
+                if (!row.is_attached())
+                    continue;
+                observers.push_back({
+                    row.get_table()->get_index_in_group(),
+                    row.get_index(),
+                    info});
+            }
+        }
+        sort(begin(observers), end(observers));
+    }
+
+    // Send didChange notifications to all observers marked as needing them
+    // Loop in reverse order to avoid O(N^2) behavior in Foundation
```

Comment:
The fact that it's O(N^2) is based on observation, not reasoning. It's not too hard to think of reasons why it would be (e.g. if `willChange:` pushes things onto the end of an array and `didChange` pops, since front popping is O(N) and back popping in O(1)), but I don't think why Foundation behaves how it does is really all that relevant.